#168064097 Fix tag content and response message

Nennyfills opened the pull request:
#### What does this PR do?
This pull request fixes the bug on tag content, response message and pagination count on following and search article endpoint.

#### Description of Task to be completed
- Rearrange tag content on search articles endpoint
- Correct response message on followings endpoint
- Add count on getting all followers and followings endpoint

#### How should this be manually tested?
- Clone this repo
Using an HTTP client test the following endpoints
- ``` GET /api/v1/:articles?```
- ``` patch /api/v1/users/profile/:userId/following```
- ```GET /api/v1/users/profile/:userId/followers```
- ```GET /api/v1/users/profile/:userId/followings```

#### Any Background Context You want to provide?
N/A

#### What are the relevant pivotal tracker stories?

[#168064097](https://pivotaltracker.com/story/show/168064097)

#### Screenshots (if appropriate)
<img width="882" alt="Screen Shot 2019-08-22 at 3 43 54 PM" src="https://user-images.githubusercontent.com/45515619/63548006-44c90f80-c525-11e9-93cb-1db059635e9e.png">
<img width="911" alt="Screen Shot 2019-08-22 at 4 05 25 PM" src="https://user-images.githubusercontent.com/45515619/63548024-4d214a80-c525-11e9-94c3-45566305d2d0.png">
<img width="878" alt="Screen Shot 2019-08-22 at 4 05 38 PM" src="https://user-images.githubusercontent.com/45515619/63548037-56aab280-c525-11e9-9c61-0fea23f3f643.png">
<img width="883" alt="Screen Shot 2019-08-22 at 4 21 16 PM" src="https://user-images.githubusercontent.com/45515619/63548064-63c7a180-c525-11e9-8f06-0291ce45132e.png">
<img width="1076" alt="Screen Shot 2019-08-22 at 7 08 56 PM" src="https://user-images.githubusercontent.com/45515619/63548084-6fb36380-c525-11e9-9653-e2ba554b14fb.png">
<img width="1036" alt="Screen Shot 2019-08-23 at 11 48 58 AM" src="https://user-images.githubusercontent.com/45515619/63588149-b8a6fe80-c59d-11e9-953f-06927d654fc6.png">
<img width="1133" alt="Screen Shot 2019-08-23 at 11 58 15 AM" src="https://user-images.githubusercontent.com/45515619/63588205-dc6a4480-c59d-11e9-8333-92f30955c9b9.png">



#### Questions:
N/A

iverenshaguy commented:
Please flatten the response of `/followees`. Please also add a screenshot of what all these responses look like with empty data

iverenshaguy commented on docs/swagger.json:
```diff
@@ -885,7 +885,7 @@
     "/profile/:userId/following": {
-      "post": {
+      "patch": {
         "description": "follower a user",
```
```suggestion
        "description": "follow a user",
```

Nennyfills replied:
Done.

iverenshaguy commented on docs/swagger.json:
```diff
@@ -885,7 +885,7 @@
     },
     "/profile/:userId/following": {
-      "post": {
+      "patch": {
```
Why was this changed?

Nennyfills replied:
I changed it because is a patch route and not a post route.

iverenshaguy replied:
What makes it a patch route?

iverenshaguy replied:
I don't really understand. How did we come to that conclusion?

iverenshaguy commented on src/helpers/genericArticleHelper.js:
```diff
@@ -5,22 +5,22 @@ const { Op } = Sequelize;
 const { Users, Categories, Articles } = models;
```
Please rename this file to something like `articleHelper.js`

Nennyfills replied:
Done.

iverenshaguy commented on src/helpers/genericArticleHelper.js:
```diff
@@ -44,35 +44,53 @@ const filter = (title, tag, author, keyword) => {
   // eslint-disable-next-line no-nested-ternary
-  const filterQuery = (keyword) ? { ...queryParams.keyword } : author ? { ...queryParams.author }
-    : { ...queryParams.query };
+  const filterQuery = keyword
```
Please refactor this piece of code to something more readable. We want to avoid nested ternary operations at all times

Nennyfills replied:
Done.

iverenshaguy commented on src/helpers/genericArticleHelper.js:
```diff
@@ -44,35 +44,53 @@ const filter = (title, tag, author, keyword) => {
+    where:
+      keyword || title || tag || author
+        ? {
+          ...filterQuery
```
```suggestion
          { ...filterQuery  }
```

Nennyfills replied:
Done.

iverenshaguy commented on src/helpers/genericArticleHelper.js:
```diff
@@ -44,35 +44,53 @@ const filter = (title, tag, author, keyword) => {
+        ? {
+          ...filterQuery
+        }
+        : undefined
```
If line 60 is not true, where would be undefined anyways so please change it to something like
```suggestion
const hasFilterQuery = keyword || title || tag || author;
...
where: hasFilterQuery && { ...filterQuery  }
...
```

Nennyfills replied:
Done.

iverenshaguy commented on src/routes/user.js:
```diff
@@ -90,7 +90,7 @@ router.get('/profile/:id', validator(profileViewSchema), asyncWrapper(verifyToke
  *
  * /profile/:userId/following:
- *   post:
+ *   patch:
```
Why was this changed?

Nennyfills replied:
I changed that part because the route is a patch route, not post route.

Nennyfills replied:
I changed it because is a patch route and not a post route.

iverenshaguy reviewed (changes requested):
Please review my comments and write tests for uncovered code

Nennyfills committed:
fix(articles and followings): fix tag content and response message

- rearrange tag content on search articles endpoint
- correct response message on followings endpoint
- add paginator count on getting all followers and followings endpoint

[Fixes #168064097]

iverenshaguy commented:
The response structure that has `following: []` should be `data`. We want to ensure consistency in our codebase

iverenshaguy commented on src/controllers/user.js:
```diff
@@ -138,9 +138,14 @@ export default {
-      data,
-      count,
-      page
+      followers: data.map((item) => {
```
followers should be data ie.e 
```suggestion
      data: whatever.map((item) => {
```

Nennyfills replied:
Done.

iverenshaguy commented on src/controllers/user.js:
```diff
@@ -168,10 +173,16 @@ export default {
       status: 'success',
-      data,
+      message: 'Request successful',
+      followings: data.map((item) => {
```
Same feedback
```suggestion
      data: whatever.map((item) => {
```

Nennyfills replied:
Done.

iverenshaguy commented on src/helpers/articleHelper.js:
```diff
@@ -5,17 +5,17 @@ const { Op } = Sequelize;
 const { Users, Categories, Articles } = models;
 
 /**
-   * filter for search article
```
So I saw that we have another file called `article` in this helpers folder. I don't really understand why the methods here were not added to that file (I'd rather they were). Is there a reason why we decided to create another file called `genericArticleHelper`? Also, I need to understand why you feel the need to use the word `generic` in the codebase. I have seen you describe functions as `generic ...` in your JSDocs. Is there a specific reason for this?

Nennyfills replied:
Before the search article PR was merged the other article file wasn't there. I called it ```genericArticleHelper``` because the sequelize operators used in the file are not limited. it can be reused anywhere to find any article or author. The function can be reused on any route to validate an empty body.

iverenshaguy replied:
Okay. Please move all methods in this file to the `./helpers/article.js` file. Also, I don't think we need to use generic. Helpers can generally be used anywhere so there's no need to specify. Thanks

iverenshaguy commented:
Please rebase and fix conflicts. Also, I have still not understood why the route was changed from post to patch.

iverenshaguy committed:
fix(articles): update implementation

- update implementation
- rebase on develop

[Fixes #168064097]
